Zend/zend_inheritance.c

```diff
@@ -1898,6 +1898,121 @@ static zend_class_entry *fixup_trait_scope(const zend_function *fn, zend_class_e
+	if (ZEND_TYPE_HAS_NAME(type)) {
+		return zend_resolve_name_type(type, ce);
+	}
+
+	/* Intersection types cannot have relative class_types */
+	if (ZEND_TYPE_IS_INTERSECTION(type)) {
+		return type;
+	}
+
+	zend_type_list *union_type_list = ZEND_TYPE_LIST(type);
+	bool has_resolved_type = false;
+	/* We don't use ZEND_TYPE_LIST_FOREACH() as we need to keep track of the array index */
+	for (uint32_t i = 0; i < union_type_list->num_types; i++) {
+		zend_type single_type = union_type_list->types[i];
+
+		ZEND_ASSERT(ZEND_TYPE_HAS_NAME(single_type) || (ZEND_TYPE_IS_INTERSECTION(single_type)));
+
+		/* Intersections types cannot have self or parent */
+		if (ZEND_TYPE_IS_INTERSECTION(single_type)) {
+			continue;
+		}
+
+		zend_type resolved_type = zend_resolve_name_type(single_type, ce);
+		if (zend_was_type_resolved(type, resolved_type)) {
+			if (!has_resolved_type) {
+				const zend_type_list *old_union_type_list = ZEND_TYPE_LIST(type);
+				union_type_list = zend_arena_alloc(&CG(arena), ZEND_TYPE_LIST_SIZE(old_union_type_list->num_types));
+				memcpy(union_type_list, old_union_type_list, ZEND_TYPE_LIST_SIZE(old_union_type_list->num_types));
+				has_resolved_type = true;
+			}
+			union_type_list->types[i] = resolved_type;
+		}
+	}
+
+	if (has_resolved_type) {
+		zend_type new_type = (zend_type) ZEND_TYPE_INIT_NONE(0);
+		ZEND_TYPE_SET_LIST(new_type, union_type_list);
+		ZEND_TYPE_FULL_MASK(new_type) |= _ZEND_TYPE_ARENA_BIT;
+		/* Inform that the type list is a union type */
+		ZEND_TYPE_FULL_MASK(new_type) |= _ZEND_TYPE_UNION_BIT;
+		return new_type;
+	} else {
+		return type;
+	}
+}
+
+static void zend_resolve_trait_relative_class_types(zend_function *const fn, const zend_class_entry *const ce)
+{
+	/* We are adding trait methods to another trait, delay resolution */
+	if (ce->ce_flags & ZEND_ACC_TRAIT) {
+		return;
+	}
+	/* No type info */
+	if (!fn->common.arg_info) {
+		return;
+	}
+
+	bool has_return_type = fn->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE;
+	/* Variadic parameters are not counted as part of the standard number of arguments */
+	bool has_variadic_type = fn->common.fn_flags & ZEND_ACC_VARIADIC;
+	uint32_t num_args = fn->common.num_args + has_variadic_type;
+	size_t allocated_size = sizeof(zend_arg_info) * (has_return_type + num_args);
+
+	zend_arg_info *new_arg_infos = fn->common.arg_info - has_return_type;
+	bool has_resolved_type = false;
+
+	for (uint32_t i = 0; i < num_args + has_return_type; i++) {
+		zend_type type = new_arg_infos[i].type;
+		zend_type resolved_type = zend_resolve_single_type(type, ce);
+		if (zend_was_type_resolved(type, resolved_type)) {
+			if (!has_resolved_type) {
+				new_arg_infos = zend_arena_alloc(&CG(arena), allocated_size);
+				memcpy(new_arg_infos, fn->common.arg_info - has_return_type, allocated_size);
+				fn->common.arg_info = new_arg_infos + has_return_type;
+				has_resolved_type = true;
+			}
+			new_arg_infos[i].type = resolved_type;
```

Comment:
Shouldn't the old type be released? I think this might only be reproducible for eval'd code where the type may contain non-interned strings.

Reply:
Maybe? Haven't thought about something like this as I would usually expect to be allocated on an arena. And releasing the type info for the traits doesn't make a lot of sense as it still needs to keep the unresolved types for itself.